hazelcast/src/main/java/com/hazelcast/internal/memory/MemoryAccessStrategy.java

```diff
@@ -0,0 +1,576 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.memory;
+
+import java.lang.reflect.Field;
+
+/**
+ * Abstract entity which provides different way to access data from external resources
+ *
+ * @param <R> - type of resource
+ */
+public interface MemoryAccessStrategy<R> extends ByteMemoryAccessStrategy<R> {
+    MemoryAccessStrategy<Object> STANDARD_MEM = MemoryAccessorStrategyProvider.getMemoryAccessStrategy(MemoryAccessStrategyType.STANDARD);
+
+    MemoryAccessStrategy<Object> STANDARD_ALIGNED_MEM = MemoryAccessorStrategyProvider.getMemoryAccessStrategy(MemoryAccessStrategyType.STANDARD_ALIGNMENT);
+
+    MemoryAccessStrategy<byte[]> HEAP_BYTE_ARRAY_MEM = MemoryAccessorStrategyProvider.getMemoryAccessStrategy(MemoryAccessStrategyType.HEAP_BYTE_ARRAY);
+
+    MemoryAccessStrategy<Object> MEM = MemoryAccessorStrategyProvider.getMemoryAccessStrategy(MemoryAccessStrategyType.PLATFORM_AWARE);
```

Comment:
We must improve these constant names. `MEM` is what `MEM` was before, i think. It should be an always-safe-to-use accessor of native CPU address space. `AMEM` is gone; it was meant to be like `MEM`, but only safe to use when the caller guarantees aligned memory access. I wonder whether we need at all the `STANDARD_ALIGNED_MEM` constant and the name is confusing (not clear whether the caller must ensure aligned access, or the accessor does it).